app/controllers/account_controller.rb

```diff
@@ -300,19 +300,22 @@ def registration_through_invitation!
     session[:auth_source_registration] = nil
 
     if @user.nil?
-      @user = User.new(language: Setting.default_language)
-    elsif user_with_placeholder_name?(@user)
-      # force user to give their name
+      @user = Users::SetAttributesService.new(
+        model: User.new, user: current_user,
+        contract_class: Users::CreateContract
+      ).call(language: Setting.default_language).result
+    else
       @user.firstname = nil
       @user.lastname = nil
     end
   end
 
   def self_registration!
     if @user.nil?
-      @user = User.new
-      @user.admin = false
-      @user.register
+      @user = Users::SetAttributesService
+        .new(model: User.new, user: current_user, contract_class: Users::CreateContract)
```

Comment:
Using the `Users::CreateContract` will lead to the validations being run which we do not want to happen here since the user did not provide any values at this point so it isn't the user's fault that the record isn't valid yet. Displaying errors right when the user opens the registration modal might be irritating:

<img width="675" alt="image" src="https://user-images.githubusercontent.com/617519/183373165-1ede7c73-efcc-4579-8a9c-4fb0fb61e88f.png">

By using the `EmptyContract` the default values are still set in the Service but validation is disabled. As long as the initialized record isn't saved, this contract can be used.